scylla/src/cloud/config.rs

```diff
@@ -468,4 +468,269 @@ mod deserialize {
+                certificateAuthorityPath: Some(good_pem_path.into()),
+                certificateAuthorityData: Some(TEST_CA.into()),
+                server: "127.0.0.1:9142".into(),
+                tlsServerName: None,
+                proxyUrl: None,
+                nodeDomain: "cql.my-cluster-id.scylla.com".into(),
+                insecureSkipTlsVerify: Some(false),
+            };
+            let dc_no_cert = super::Datacenter {
+                certificateAuthorityPath: None,
+                certificateAuthorityData: None,
+                ..dc_valid.clone()
+            };
+            let dc_invalid_cert = super::Datacenter {
+                certificateAuthorityData: Some("INVALID CERFITICATE".into()),
+                ..dc_valid.clone()
+            };
+            let dc_cert_nonfound = super::Datacenter {
+                certificateAuthorityPath: Some(no_pem_path.into()),
+                certificateAuthorityData: None,
+                ..dc_valid.clone()
+            };
+            let dc_cert_found_bad = super::Datacenter {
+                certificateAuthorityPath: Some(bad_pem_path.into()),
+                certificateAuthorityData: None,
+                ..dc_valid.clone()
+            };
+            let dc_cert_found_good = super::Datacenter {
+                certificateAuthorityPath: Some(good_pem_path.into()),
+                certificateAuthorityData: None,
+                ..dc_valid.clone()
+            };
+            let dc_bad_domain_empty = super::Datacenter {
+                nodeDomain: "".into(),
+                ..dc_valid.clone()
+            };
+            let dc_bad_domain_trailing_minus = super::Datacenter {
+                nodeDomain: "cql.scylla-.com".into(),
+                ..dc_valid.clone()
+            };
+            let dc_good_domain_interior_minus = super::Datacenter {
+                nodeDomain: "cql.scylla-cloud.com".into(),
+                ..dc_valid.clone()
+            };
+            let dc_bad_domain_special_sign = super::Datacenter {
+                nodeDomain: "cql.$cylla-cloud.com".into(),
+                ..dc_valid.clone()
+            };
+            let dc_bad_server_not_url = super::Datacenter {
+                server: "NotAUrl".into(),
+                ..dc_valid.clone()
+            };
+
+            let good = [dc_valid, dc_cert_found_good, dc_good_domain_interior_minus];
+
+            let bad = [
+                dc_no_cert,
+                dc_invalid_cert,
+                dc_cert_nonfound,
+                dc_cert_found_bad,
+                dc_bad_domain_empty,
+                dc_bad_domain_special_sign,
+                dc_bad_domain_trailing_minus,
+                dc_bad_server_not_url,
+            ];
+
+            for test in good {
+                super::super::Datacenter::try_from(test).unwrap();
+            }
+            for test in bad {
+                super::super::Datacenter::try_from(test).unwrap_err();
+            }
```

Comment:
Isn't it prevalent in our tests to stop on the first failed case?

One way I could transform the tests without much hassle is by using distinct named `#[test]` function for each test - this would give us reporting for specific cases for free.

Reply:
I've made all those tests separate.

Reply:
In principle, it's good to have separate test cases for each thing you want to check. If you break something then having multiple small tests failing gives you better information about what exactly you have broken (especially if multiple tests fail).

Of course, it's not always practical or easy to split tests like this, but here it shouldn't be too hard.